src/middlewared/middlewared/plugins/truesearch.py

```diff
@@ -0,0 +1,170 @@
+import asyncio
+import os
+
+from middlewared.service import Service
+from middlewared.utils.smb import SearchProtocol
+
+
+class TrueSearchService(Service):
+    class Config:
+        private = True
+
+    async def available(self) -> bool:
+        """
+        Whether the truesearch service can be used.
+        """
+        # Boot pool is usually too small to fit the search index
+        return not await self.middleware.call('systemdataset.is_boot_pool')
+
+    async def enabled(self) -> bool:
+        """
+        Whether the truesearch service should be started.
+        """
+        return await self.available() and bool(await self.directories())
+
+    async def directories(self) -> list[str]:
+        """
+        What directories will it index.
+        """
+        return await self.process_directories(await self.raw_directories())
+
+    async def process_directories(self, directories: set[str]) -> list[str]:
+        """
+        :param directories: a list of share directories
+        :return: a list of directories that TrueSearch will index
+
+        The reason these do not match is that TrueSearch does not traverse filesystem boundaries.
+        So if a share directory includes a nested dataset, it must be explicitly included in the directories list.
+        Encrypted datasets are excluded from the index to prevent sensitive data being stored in index unencrypted.
+        """
+        pools = {directory.removeprefix('/mnt/').split('/')[0] for directory in directories}
+        mountpoints = {
```

Comment:
This dict comprehension is miserable. Please do the old fashioned way of something like
```
mount points = {}
for i in await...
    if i['type'] != 'FILESYSTEM':
        continue
    if I['properties']['mountpoint']['value']
```
Also why are you checking to see if `mountpoint` startswith `/mnt/`? Are you trying to determined if it's mounted or something?